kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -0,0 +1,662 @@
+public class TableFeatures {
+
+  /////////////////////////////////////////////////////////////////////////////////
+  /// START: Define the {@link TableFeature}s                                   ///
+  /// If feature instance variable ends with                                    ///
+  ///  1) `_W_FEATURE` it is a writer only feature.                             ///
+  ///  2) `_RW_FEATURE` it is a reader-writer feature.                          ///
+  /////////////////////////////////////////////////////////////////////////////////
+  public static final TableFeature APPEND_ONLY_W_FEATURE = new AppendOnlyFeature();
+
+  private static class AppendOnlyFeature extends TableFeature.LegacyWriterFeature {
+    AppendOnlyFeature() {
+      super(/* featureName = */ "appendOnly", /* minWriterVersion = */ 2);
+    }
+
+    @Override
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return TableConfig.APPEND_ONLY_ENABLED.fromMetadata(metadata);
+    }
+
+    @Override
+    public boolean hasKernelWriteSupport(Metadata metadata) {
+      return true;
+    }
+  }
+
+  public static final TableFeature INVARIANTS_W_FEATURE = new InvariantsFeature();
+
+  private static class InvariantsFeature extends TableFeature.LegacyWriterFeature {
+    InvariantsFeature() {
+      super(/* featureName = */ "invariants", /* minWriterVersion = */ 2);
+    }
+
+    @Override
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return hasInvariants(metadata.getSchema());
+    }
+
+    @Override
+    public boolean hasKernelWriteSupport(Metadata metadata) {
+      // If there is no invariant, then the table is supported
+      return !hasInvariants(metadata.getSchema());
+    }
+  }
+
+  public static final TableFeature CONSTRAINTS_W_FEATURE = new ConstraintsFeature();
+
+  private static class ConstraintsFeature extends TableFeature.LegacyWriterFeature {
+    ConstraintsFeature() {
+      super("constraints", /* minWriterVersion = */ 3);
+    }
+
+    @Override
+    public boolean hasKernelWriteSupport(Metadata metadata) {
+      // Kernel doesn't support table with constraints.
+      return !hasCheckConstraints(metadata);
+    }
+
+    @Override
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return hasCheckConstraints(metadata);
+    }
+  }
+
+  public static final TableFeature COLUMN_MAPPING_RW_FEATURE = new ColumnMappingFeature();
+
+  private static class ColumnMappingFeature extends TableFeature.LegacyReaderWriterFeature {
+    ColumnMappingFeature() {
+      super("columnMapping", /* minReaderVersion = */ 2, /* minWriterVersion = */ 5);
+    }
+
+    @Override
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return TableConfig.COLUMN_MAPPING_MODE.fromMetadata(metadata) != NONE;
+    }
+
+    @Override
+    public boolean hasKernelReadSupport() {
```

Comment:
Q: why override this method to return `true` if `TableFeature:: hasKernelReadSupport` already returns true by default?